126 create continuousdata class

psomhorst opened the pull request.

psomhorst commented:
This PR has become much more than creating a continuousdata class.

This PR now:
- removes Variants, VariantCollections
- introduces DataCollections
- introduces ContinuousData and SparseData
- updates Sequence to contain a DataCollection for each of EITData, ContinuousData and SparseData

psomhorst commented:
@DaniBodor Thanks :)

I'll get to documentation this week, and will look into making methods private.

DaniBodor commented on eitprocessing/eit_data/timpel.py:
```diff
@@ -88,43 +88,76 @@ def _from_path(  # pylint: disable=too-many-arguments,too-many-locals
-            phases.append(MinValue(index, time[int(index)]))
+            phases.append(MinValue(index, time[int(index)]))  # noqa: PERF401
 
         for index in np.flatnonzero(data[:, 1028] == 1):
-            phases.append(MaxValue(index, time[int(index)]))
+            phases.append(MaxValue(index, time[int(index)]))  # noqa: PERF401
 
         for index in np.flatnonzero(data[:, 1029] == 1):
-            phases.append(QRSMark(index, time[int(index)]))
+            phases.append(QRSMark(index, time[int(index)]))  # noqa: PERF401
```
shouldn't these be loaded as SparseData? Or are we not handling those as such yet?

DaniBodor commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -144,71 +148,75 @@ def _ensure_path_list(path: PathArg) -> list[Path]:
-        from .draeger import DraegerEITData  # pylint: disable=import-outside-toplevel
-        from .sentec import SentecEITData  # pylint: disable=import-outside-toplevel
-        from .timpel import TimpelEITData  # pylint: disable=import-outside-toplevel
+        from .draeger import DraegerEITData
+        from .sentec import SentecEITData
+        from .timpel import TimpelEITData
```
Why are these imported outside of top level?

DaniBodor reviewed (changes requested):
Really a lot got done here.
Please see my inline comments for a few comments and suggestions, but overall it's looking really good and it seems that when these are done, we finally will have a working package again that we can start adding to rather than fixing :)

I haven't gone into fixing all the linting errors yet. I think maybe after this PR is done and merged, I will do a separate PR to fix all of that before we start a new big one. I can probably do that within a day of merging this one, so that there isn't much delay with carrying on after this. That would at least be my preference over adding it onto this, what do you think?

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -144,71 +148,75 @@ def _ensure_path_list(path: PathArg) -> list[Path]:
-        if np.min(b.time) <= np.max(a.time):
-            raise ValueError(f"{b} (b) starts before {a} (a) ends.")
-        time = np.concatenate((a.time, b.time))
+        if np.min(other.time) <= np.max(self.time):
+            msg = f"{other} (b) starts before {self} (a) ends."
+            raise ValueError(msg)
+        time = np.concatenate((self.time, other.time))
```
I agree. I've been working the addition mixin. I just pushed 129c_addition_mixin_psomhorst* that solves this issue by adding a flag `check_time_consistency` that allows for overruling this check.

* I created a new branch instead of pushing to your branch, because it's rebased on this PR.

psomhorst commented on eitprocessing/continuous_data/__init__.py:
```diff
@@ -1,54 +1,82 @@
+        description: Human readible extended description of the data.
+        parameters: ???
+        loaded: True if raw data was loaded directly from source. False if the data was derived.
+        derived_from: Traceback of intermediates from which the current data was derived.
```
Agreed. I changed it to a property, that return True if the length of `derived_from` equals 0 in 15b424e.

psomhorst commented on eitprocessing/continuous_data/__init__.py:
```diff
@@ -1,54 +1,82 @@
+            name=name or label,
+            unit=unit or self.unit,
+            description=description or f"Derived from {self.name}",
+            parameters=self.parameters | (parameters or {}),
```
`|`  merges two dictionaries.

"Create a new dictionary with the merged keys and values of d and other, which must both be dictionaries. The values of other take priority when d and other share keys." from https://docs.python.org/3.12/library/stdtypes.html#dict.

This results in a merge of self.parameters and the provided parameters. If parameters is None, self.parameters is merged with {}, resulting in a copy of self.parameters.

DaniBodor replied:
Thanks for clarifying. I misunderstood this before :)

psomhorst commented on eitprocessing/continuous_data/__init__.py:
```diff
@@ -1,54 +1,82 @@
+
+@dataclass
+class ContinuousData:
+    """Data class for (non-EIT) data with a continuous time axis.
```
Yes!

Fixed in f71d69c170cd99e91550d5197cba1bc3581cf439 and 224a964939e0cdfb723750e2989a98715dd0f51a.

psomhorst commented on eitprocessing/data_collection/__init__.py:
```diff
@@ -0,0 +1,66 @@
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+V = TypeVar("V")
```
Agreed.

I added this in db055eff12512a442976aa4b31d6885fa5d1536d. The __init__() method now also warns the used when passing a different type than either of those.

psomhorst commented on eitprocessing/data_collection/__init__.py:
```diff
@@ -0,0 +1,66 @@
+
+        if key and key != item.label:
+            msg = f"'{key}' does not match label '{item.label}'."
+            raise KeyError(msg)
```
I added comments in dc591a409f57d207c5ecd3f77880a1aa70396a34.

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -115,26 +108,37 @@ def from_path(  # pylint: disable=too-many-arguments,too-many-locals
-                assert isinstance(eit, EITData)
-                assert isinstance(continuous, ContinuousDataCollection)
-                assert isinstance(sparse, SparseDataCollection)
+                assert isinstance(eit, DataCollection)  # noqa: S101
+                assert isinstance(continuous, DataCollection)  # noqa: S101
+                assert isinstance(sparse, DataCollection)  # noqa: S101
```
I changed it to DataCollection([...]Data) instead in 0ee24d6001576f3ec7567ca2a479fa084ba5a5d5.

psomhorst replied:
These lines are here purely for type checking. In b3abe8cee793b46ad7bcbc9359a62706bcc397aa I put the checks within a `if TYPE_CHECKING` block, removing the need for the #noqa, and added a comment explaining.

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -217,78 +225,76 @@ def _sliced_copy(
 
 
 @dataclass(eq=False)
-class EITData_(EITData):
+class EITData_(EITData):  # noqa: N801
     vendor: Vendor = field(init=False)
```
Yes.

`EITData.from_path()` requires `vendor` as argument, so it can choose the right class to load the file. 
`DraegerEITData.from_path()` does not allow `vendor` as argument, because the vendor is implicit in the choice of class. Of course the same is true for SentecEITData and TimpelEITData.
It is possible to override `from_path()` in each of the three subclasses, but that would not be DRY. 

`EITData_` removes `vendor` from the initializer with `field(init=False)`. 
Each subclass sets `vendor` to the proper value.
`EITData_` also changes the arguments required for `from_path()`.

I wasn't able to solve this without an intermediate class. Any other solution I came up with requires re-implementing `from_path()` in each of the three subclasses.

psomhorst replied:
I think EITData_ might better be named _EITData, to indicate it's for internal use only (rather than just a similarly-named copy of EITData).

DaniBodor replied:
Thanks for the explanation. Agreed about the second comment. Let's leave it like this for now, and I will ask around if someone in the office has a more elegant solution to this.

DaniBodor replied:
Opened #152 for this

DaniBodor commented:
@psomhorst , great that you are picking this up so quickly. Please tag me when you want me to look at/re-review the changes. Until then I will not track what is happening on this PR :)

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -79,27 +74,25 @@ def from_path(  # pylint: disable=too-many-arguments,too-many-locals
+        sparse_datasets: list[DataCollection] = []
 
         for single_path in paths:
             # this checks whether each path exists before any path is loaded to
             # prevent unneccesary loading
             single_path.resolve(strict=True)  # raises if file does not exists
 
         for single_path in paths:
-            loaded_data = vendor_class._from_path(  # pylint: disable=protected-access
+            loaded_data = vendor_class._from_path(  # noqa: SLF001
```
`_ensure_path_list()` has the task of wrapping a single path in a list. I would not want to abuse that function to perform checks. We could write another method instead.

I certainly would want this to raise an exception if any file is unavailable. If you try to load a file that does not exist, somehas has gone very wrong. In the GUI, you select files just before loading. In analysis scripts, you want to be sure you load what you think you load. If a script starts working with a subset of that data, a lot of unexpected things could happen.

`cls._from_path()` is an abstract class when calling `EITData.from_path(path, vendor="Draeger")`. The specific vendor class has to be selected first.

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -35,27 +32,24 @@ class EITData(SelectByTime, Equivalence, ABC):
         max_frames: int | None = None,
         return_non_eit_data: bool = False,
-    ) -> Self | tuple[Self, ContinuousDataCollection, SparseDataCollection]:
-        """Load sequence from path(s)
+    ) -> DataCollection | tuple[DataCollection, DataCollection, DataCollection]:
+        """Load sequence from path(s).
```
Agreed. 

`from_path()` returns either a DataCollection(EITData), or a tuple containing DataCollection(EITData), DataCollection(ContinuousData), DataCollection(SparseData). If there is no sparse data, the third DataCollection has no items. 

This whole way of loading data is bothering me. It adds a lot of complexity (as you rightly point out). I think a good solution would be to separate loading data from the classes, as is common practice. But this would require yet another rewrite and changes to the GUI. It is something to consider/discuss next week?

DaniBodor replied:
How do you feel about naming this argument `attach_additional_data` or `attach_external_data` or something like that?

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -35,27 +32,24 @@ class EITData(SelectByTime, Equivalence, ABC):
+    def from_path(
         cls,
         path: PathArg,
         vendor: Vendor | str,
```
No, I guess you're correct. I added this because there were different type declarations which were inconsistent with each other. I removed the custom aliases in 8adfceac4c5912da941e84e9a12dbb36c7469f92.

psomhorst commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -115,26 +108,37 @@ def from_path(  # pylint: disable=too-many-arguments,too-many-locals
-        return reduce(cls.concatenate, eit_datasets)
+        return reduce(DataCollection.concatenate, eit_datasets)
+
+    @classmethod
+    @abstractmethod
+    def _from_path(
+        cls: type[Self],
+        path: Path,
+        framerate: float | None = None,
+        first_frame: int | None = None,
+        max_frames: int | None = None,
+        return_non_eit_data: bool = False,
+    ) -> DataCollection | tuple[DataCollection, DataCollection, DataCollection]:
+        ...
```
See my earlier comment. Let's discuss whether to revisit how to load data, and change accordingly.

DaniBodor replied:
Also referenced in #152

psomhorst commented on eitprocessing/eit_data/sentec.py:
```diff
@@ -187,13 +183,12 @@ def _read_frame(  # pylint: disable=too-many-arguments
                 f"height ({image_height}) of the frame."
-                f"Image will not be stored"
+                f"Image will not be stored",
             )
```
Probably. I guess this was copied from the Matlab code. I've never encountered this warning. I guess it is not common to come across malformed data.

I changed it to an OSError in af44e2bace3d4c5c691e7be5afee324a400cab91.

psomhorst commented:
Hi @DaniBodor. Thanks for the review! Great comments overall.

I've addressed all your comments. Most I agree on and have made changes.

You have noticed the big issue with EITData: data loading complicates a lot of things. We could consider splitting functionality and data loading (as is common practise), but I'm hesitant to start yet another rewrite.

DaniBodor commented on eitprocessing/data_collection/__init__.py:
```diff
@@ -1,19 +1,25 @@
     data_type: type
 
     def __init__(self, data_type: type[V], *args, **kwargs):
+        if data_type not in (EITData, ContinuousData, SparseData):
+            warnings.warn(f"Type {data_type} not expected to be stored in a DataCollection.")
```
In saw the docstring commit after this, where you already comment on other data types. Not sure we should allow that at all (even if we mention it's not supported). In my mind, it would be a rare occurrence, but is asking for problems if someone would try to do it.

DaniBodor commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -109,10 +109,12 @@ def from_path(
-                assert isinstance(eit, DataCollection)  # noqa: S101
-                assert isinstance(continuous, DataCollection)  # noqa: S101
-                assert isinstance(sparse, DataCollection)  # noqa: S101
+                if TYPE_CHECKING:
+                    # These asserts tells the type checker they can assume these are the types of these variables.
+                    # Required due to multiple return types for _from_path().
+                    assert isinstance(eit, DataCollection)
+                    assert isinstance(continuous, DataCollection)
+                    assert isinstance(sparse, DataCollection)
```
If it's just for type checking, then I believe you can also do this:

```suggestion
                eit: DataCollection
                continuous: DataCollection
                sparse: DataCollection
```

psomhorst replied:
Fixed this in 695c2abb1b72d0fe57f5a35b531fdaa9968c3a4f.

DaniBodor commented on eitprocessing/eit_data/__init__.py:
```diff
@@ -217,78 +225,76 @@ def _sliced_copy(
-        label: str | None = None,
-        framerate: float | None = None,
-        first_frame: int | None = None,
-        max_frames: int | None = None,
-        return_non_eit_data: bool = False,
-    ) -> Self | tuple[Self, ContinuousDataCollection, SparseDataCollection]:
-        ...
+    @property
+    def pixel_impedance_global_offset(self) -> np.ndarray:
+        return self.pixel_impedance - self.global_baseline
+
+    @property
+    def pixel_baseline(self) -> np.ndarray:
+        return np.nanmin(self.pixel_impedance, axis=0)
+
+    @property
+    def pixel_impedance_individual_offset(self) -> np.ndarray:
+        return self.pixel_impedance - np.min(self.pixel_impedance, axis=0)
+
+    @property
+    def global_impedance(self) -> np.ndarray:
+        return np.nansum(self.pixel_impedance, axis=(1, 2))
```
shall we then also use our meeting to discuss these names?

DaniBodor reviewed (commented):
Thanks for all the explanations and edits @psomhorst. I've responded to a few comments and left a few more inline comments as well.

psomhorst commented on eitprocessing/continuous_data/__init__.py:
```diff
@@ -117,14 +123,14 @@ def lock(self) -> None:
         The values can be unlocked using `unlock()`.
         """
-        self.values.flags["WRITEABLE"] = False
+        getattr(self, attr).flags["WRITEABLE"] = False
```
I agree. I added this in 1b0b10f5baeb4f3642b404d9a4e7681c88d21f4b with `*attr`.

psomhorst commented on eitprocessing/data_collection/__init__.py:
```diff
@@ -1,19 +1,25 @@
     from typing_extensions import Self
 
-V = TypeVar("V")
+V = TypeVar("V", bound=(EITData, ContinuousData, SparseData))
```
You are correct. I was confused by your use of the word bound, while it should be a constraint.

psomhorst commented on eitprocessing/data_collection/__init__.py:
```diff
@@ -1,19 +1,25 @@
+from eitprocessing.continuous_data import ContinuousData
+from eitprocessing.eit_data import EITData
 from eitprocessing.mixins.equality import Equivalence
+from eitprocessing.sparse_data import SparseData
```
I think a87c671ac6c81624967355ff545e69984012334b fixes it.

DaniBodor replied:
I still get circular imports (which i don't if I revert to not having any constraints on `V`). The commit you refer to touches another module than this (`eit_data/__init__.py` rather than `data_collection/__init__.py`)

psomhorst replied:
I missed that one. I think it is now solved in b993ef2774100f330c663360e984fc9c12344af9. It's not pretty, but this is the result of interdependence between classes.

psomhorst commented:
@DaniBodor Thanks for the added comments. 

For clarity I marked all conversations that I think are resolved as such. There are only a few comments I replied to still open.

DaniBodor commented:
@psomhorst , I added a few changes to docstrings, take a look and see if you agree.
Apart from that, I do still get [circular import](https://github.com/EIT-ALIVE/eitprocessing/pull/142#discussion_r1514542412) problems if I run pytest. Do you see a way around that, apart from reverting to unconstrained `V`?
